### ISSUE TITLE
PHD is never disconnecting the telescope with ASCOM

### ISSUE THREAD

**Issue by kevinferrare** [OpenPHDGuiding/phd2#624]
I can see here that the connected property of the ASCOM driver is never set to false when clicking on disconnect:
https://github.com/OpenPHDGuiding/phd2/blob/master/scope_ascom.cpp#L416

This can cause issues in some drivers:
 - Reconnection will not work because the resource are not released
 - As the driver receives no disconnection signal, mount movement could continue

**Comment by agalasso** [comment]
Hi Kevin, PHD2 releases the driver COM object reference when it disconnects from the scope.  An ASCOM driver _must_ handle the case of a client disconnecting without explicitly setting the Connected property to false. If the driver relies on clients setting Connected = false, then the driver is broken... consider the case of an application crashing or being killed with Task manager. In that scenario the COM object reference is released but Disconnected is never set to false.
You can see some discussion about this in the PHD2 forum, like this discussion thread here: https://groups.google.com/d/msg/open-phd-guiding/ZGJF8oKj4SQ/MaJIbJbIBwAJ
Andy

**Comment by kevinferrare** [comment]
Hi @agalasso, and thanks for the clarification.
Before opening this issue, I tried several things, but I was never able to detect within the driver that the object reference is released (Finalize() and Dispose() methods are never called when clicking on disconnect in PHD).
Do you have any clue on how to catch this within the driver?

**Comment by agalasso** [comment]
sure, let's take this discussion to email: andy.galasso@gmail.com

**Comment by bwdev01** [comment]
Hi Kevin.  I know you've taken this off-forum with Andy, which is good.  If
you've written your own ASCOM driver, I hope you followed one of the two
templates and didn't just roll your own.  If you used the LocalServer
option, the local server code does its own reference-counting because your
code will see only one client and can't know how many clients there are.
When the last client is disconnected, your driver instance will be disposed
by the LocalServer code.   All of this is tied up in a lot of COM-interop
weirdness, which is why I found it best to use the ASCOM template.  Also,
the 'connected' property turns out to be somewhat ambiguous in the ASCOM
spec - does it mean the client is disconnecting from the driver or the
driver from the mount?  Most people assume the former, but there are some
drivers that handle a 'connected=false' op as a trigger to disconnect from
the mount.  For those drivers, it would be a mess for PHD2 to set the
connected property to false because it would blow all the other clients out
of the water.

 

Hope this helps,

Bruce

 

 

  _____

**Comment by kevinferrare** [comment]
Hi Bruce, thanks for the insights!

Yes, I used the template "ASCOM Device Driver (C#)".
LocalServer did not seem to make sense for me when I created the driver since:
- The hardware in my case only controls the mount hardware.
- There is already the "POTH Hub (ASCOM)" when several client connection to a single driver is needed.

However, I may be misunderstanding the purpose of LocalServer, since there is very few documentation available on this ...

Regarding the "connected" property, the documentation says "Set True to connect to the device hardware. Set False to disconnect from the device hardware.":
http://www.ascom-standards.org/Help/Developer/html/P_ASCOM_DriverAccess_AscomDriver_Connected.htm
So I followed this when developping the driver ... I may miss something, but it didn't seem very ambiguous to me ...

Something that I don't get is whether it is the responsibility of the driver to handle multiple client connections or not, I thought that the POTH Hub was developed to help drivers focus on the task of interfacing with the hardware.

Kind regards,
Kevin
